server/src/main/java/org/elasticsearch/tracing/Traceable.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tracing;
+
+import java.util.Map;
+
+/**
+ * Something which maps onto a <i>span</i> in a distributed trace.
+ */
+public interface Traceable {
```

Comment:
The abstraction was introduced in the original SpaceTime project. TBH I never thought about changing it.

We have this abstraction because we won't stop with tracing tasks. For example, in https://github.com/pugnascotia/elasticsearch/commit/06d6105ed10bc51fd8b507c34f2969e976e5c445 (and bear in mind that is just a quickly hacked-together example, hence it does some questionable things), we're tracing `FetchSubPhaseProcessor` instances, and there's quite a few implementations of that interface. It's not difficult to imagine that each span might want to add different attributes to the span, and it would be better for that logic to live on the individual implementations that rather in the `FetchPhase` class.

It's also important the the logic for the span ID is the same when opened and closing spans, so having it in once place via the interface ensures this.

Having said all that, I would be particularly interested for us to add tracing to e.g. repository API requests (S3), so I would be very interested to know whether you think this API could work for that, or if we need to make some adjustments. Would you be able to do that?

Reply:
> It's also important the the logic for the span ID is the same when opened and closing spans, so having it in once place via the interface ensures this.

But it's not really "one place". Now we're forcing new implicit life-cycle requirements on things like the rest-channel (which to me seems the worst one from this perspective) and making our whole codebase less flexible because any change to functionality that implements this interface now has to account for the tracing logic.

If you build the three objects here in-line where they're actually used you avoid complicating the codebase like that and we retain all flexibility. For the specific case of the fetch phase we could still add an abstraction to cover different implementations maybe but for all the other spots forcing this interface on top of random existing classes will increase the cost of future changes needlessly as far as I can tell.
If you think it's easiest to piggy back onto existing classes at this point, a better solution to this may be to just have `RestChannel` or `Task` return a `Traceable` via a getter without introducing a common abstraction across these very different classes. Then if we want to change or remove any of these we can just pull the `Traceable` out of them and pass it around some other way pretty easily? (also doing this `Traceable` could just become a `record` and doesn't even need to be an interface maybe because we don't have any state associated with an instance anyway as its getters only?)